db/version_set.cc

```diff
@@ -5831,10 +5834,22 @@ InternalIterator* VersionSet::MakeInputIterator(
       if (c->level(which) == 0) {
         const LevelFilesBrief* flevel = c->input_levels(which);
         for (size_t i = 0; i < flevel->num_files; i++) {
+          const FileMetaData& fmd = *flevel->files[i].file_metadata;
+          if (start.has_value() &&
+              cfd->user_comparator()->Compare(start.value(),
+                                              fmd.largest.user_key()) > 0) {
+            continue;
+          }
+          if (end.has_value() &&
+              cfd->user_comparator()->Compare(end.value(),
+                                              fmd.smallest.user_key()) < 0) {
```

Comment:
```suggestion
              cfd->user_comparator()->Compare(end.value(),
                                              fmd.smallest.user_key()) <= 0) {
```
I think we should also skip when the end key equals to the smallest key, as the end is exclusive:
https://github.com/facebook/rocksdb/blob/4bb143eb6514e6361880608c1c0e07f7c0e24d3a/db/compaction/compaction_job.cc#L130-L131

As the boundaries are either the smallest key or the largest key. In your test example, the subcompaction ranges are:
```
   sub_comp1.      sub_comp2.          sub_comp3.
[*, key000397] [key000397, key000799] [key000799, *]
```
both `sub_comp1` and `sub_comp2` includes an extra file which is not needed. Still much better than without the PR, but I think it might worth to skip that too.

Reply:
I'm a little bit nervous about it, as the boundary key is always prone to bugs. I would be a little bit more conservative about it. Let me add a comment here.

Reply:
I added a comment. Feel free to optimize it later when you have courage to do so:)